File: app/src/main/java/com/juawapps/newsspread/data/api/adapters/UnwrapConverterFactory.java
package com.juawapps.newsspread.data.api.adapters;

import java.lang.annotation.Annotation;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.Converter;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by joaocevada on 14/12/2017.
 */

public class UnwrapConverterFactory extends Converter.Factory {

    private final GsonConverterFactory factory;

    public UnwrapConverterFactory(GsonConverterFactory factory) {
        this.factory = factory;
    }

    @Override
    public Converter<ResponseBody, ?> responseBodyConverter(final Type type,
                                                            Annotation[] annotations, Retrofit retrofit) {
        Type wrappedType = new ParameterizedType() {
            @Override
            public Type[] getActualTypeArguments() {
                return new Type[]{type};
            }

            @Override
            public Type getOwnerType() {
                return null;
            }

            @Override
            public Type getRawType() {
                return WrappedApiResponse.class;
            }
        };

        Converter<ResponseBody, ?> gsonConverter = factory
                .responseBodyConverter(wrappedType, annotations, retrofit);

        return new UnwrapResponseConverter(gsonConverter);
    }

    @Override
    public Converter<?, RequestBody> requestBodyConverter(Type type,
                                                          Annotation[] parameterAnnotations, Annotation[] methodAnnotations, Retrofit retrofit) {
        return factory
                .requestBodyConverter(type, parameterAnnotations, methodAnnotations, retrofit);
    }

}

File: app/src/main/java/com/juawapps/newsspread/data/api/adapters/UnwrapConverterFactory.kt
package com.juawapps.newsspread.data.api.adapters

import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

import okhttp3.RequestBody
import okhttp3.ResponseBody
import retrofit2.Converter
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

/**
 * Factory to unwrap an api response to an actual object without the {status="status", content="content"} object.
 */

class UnwrapConverterFactory(private val factory: GsonConverterFactory) : Converter.Factory() {

    override fun responseBodyConverter(type: Type,
                                       annotations: Array<Annotation>, retrofit: Retrofit): Converter<ResponseBody, *>? {

        val wrappedType = object : ParameterizedType {
            override fun getActualTypeArguments(): Array<Type> {
                return arrayOf(type)
            }

            override fun getOwnerType(): Type? {
                return null
            }

            override fun getRawType(): Type {
                return WrappedApiResponse::class.java
            }
        }

        val gsonConverter: Converter<ResponseBody?, *>? = factory
                .responseBodyConverter(wrappedType, annotations, retrofit)


        return UnwrapResponseConverter(gsonConverter)
    }

    override fun requestBodyConverter(type: Type?,
                                      parameterAnnotations: Array<Annotation>, methodAnnotations: Array<Annotation>, retrofit: Retrofit): Converter<*, RequestBody>? {
        return factory
                .requestBodyConverter(type!!, parameterAnnotations, methodAnnotations, retrofit)
    }

}
